tests/portForwardingServer.spec.ts

```diff
@@ -0,0 +1,147 @@
+    } as ConnectOptions);
+    const page = await browser.newPage();
+    await page.goto(server.PREFIX + '/foo.html');
+    expect(await page.content()).toContain('original-target');
+    expect(reachedOriginalTarget).toBe(true);
+    await browserServer.close();
+  });
+
+  it('should proxy local requests', async ({ browserType, browserOptions, server }, workerInfo) => {
+    process.env.PW_TEST_PROXY_TARGET = port.toString();
+    let reachedOriginalTarget = false;
+    server.setRoute('/foo.html', async (req, res) => {
+      reachedOriginalTarget = true;
+      res.end('<html><body></body></html>');
+    });
+    const examplePort = 20_000 + workerInfo.workerIndex * 3;
+    const browserServer = await browserType.launchServer({
+      ...browserOptions,
+      _acceptForwardedPorts: true
+    } as LaunchOptions);
+    const browser = await browserType.connect({
+      wsEndpoint: browserServer.wsEndpoint(),
+      _forwardPorts: [examplePort]
+    } as ConnectOptions);
+    const page = await browser.newPage();
+    await page.goto(`http://localhost:${examplePort}/foo.html`);
+    expect(await page.content()).toContain('from-retargeted-server');
+    expect(reachedOriginalTarget).toBe(false);
+    await browserServer.close();
+  });
+
+  it('should lead to the error page for forwarded requests when the connection is refused', async ({ browserType, browserOptions, browserName, isWindows}, workerInfo) => {
+    const examplePort = 20_000 + workerInfo.workerIndex * 3;
+    const browserServer = await browserType.launchServer({
+      ...browserOptions,
+      _acceptForwardedPorts: true
+    } as LaunchOptions);
+    const browser = await browserType.connect({
+      wsEndpoint: browserServer.wsEndpoint(),
+      _forwardPorts: [examplePort]
+    } as ConnectOptions);
+    const page = await browser.newPage();
+    const response = await page.goto(`http://localhost:${examplePort}`);
+    expect(response.status()).toBe(502);
+    await page.waitForSelector('text=Could not connect');
+    await browserServer.close();
+  });
+
+  it('should lead to the error page for non-forwarded requests when the connection is refused', async ({ browserName, browserType, browserOptions, isWindows}, workerInfo) => {
+    process.env.PW_TEST_PROXY_TARGET = '50001';
+    const browserServer = await browserType.launchServer({
+      ...browserOptions,
+      _acceptForwardedPorts: true
+    } as LaunchOptions);
+    const browser = await browserType.connect({
+      wsEndpoint: browserServer.wsEndpoint(),
+      _forwardPorts: []
+    } as ConnectOptions);
+    const page = await browser.newPage();
+    const response = await page.goto(`http://localhost:44123/non-existing-url`);
+    expect(response.status()).toBe(502);
+    await page.waitForSelector('text=Could not connect');
+
+    await browserServer.close();
+  });
+
+  it('should not allow connecting a second client when _acceptForwardedPorts is used', async ({ browserType, browserOptions }, workerInfo) => {
```

Comment:
Let's add a test that throws when _acceptForwardedPorts is false.

Reply:
good idea, added.